### PR TITLE
Provide access to the name of contracts.

### PR DESCRIPTION
**Pull request by axic**
Part of #1447.

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1361,6 +1361,17 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 			);
 			m_context << Instruction::POP;
 		}
+		else if (member == "name")
+		{
+			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
+			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			utils().fetchFreeMemoryPointer();
+			// store string length
+			m_context << Instruction::DUP1 << u256(contract.name().length()) << Instruction::MSTORE;
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Shouldn't the arguments be reversed?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1361,6 +1361,17 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
...
+			// adjust pointer
+			m_context << Instruction::DUP1 << u256(32) << Instruction::ADD;
+			utils().storeStringData(contract.name());
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1374]: What about updating the free memory pointer?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 3]
```diff
@@ -15113,6 +15113,35 @@ BOOST_AUTO_TEST_CASE(code_access_content)
...
+				return type(D).name;
+			}
+			function name2() public pure returns (string memory) {
+				return type(C).name;
+			}
+		}
+	)";
+
+	compileAndRun(sourceCode, 0, "C");
+	bytes argsC = encodeArgs(u256(0x20), u256(1), "C");
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:15146]: Please also test names longer than 32 bytes.

**Comment by chriseth** [comment]
Please add a test where the name of a contract is assigned to a constant.

**Comment by axic** [comment]
> Please add a test where the name of a contract is assigned to a constant.

You mean something like
```
contract C {
  string public constant name = type(C).name;
```

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -1361,6 +1361,18 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 			);
 			m_context << Instruction::POP;
 		}
+		else if (member == "name")
+		{
+			TypePointer arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
+			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
+			m_contract << u256(contract.name().length() + 32);
+			utils().allocateMemory();
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp:1369]: How about an overload where `allocateMemory` takes an `unsigned`?

I see in many cases the pattern in
```
m_context << number;
allocateMemory():
```

With the overload we could avoid the `SWAP` at this stage.

Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1369]: Sure!

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add a test with an assignment to a constant state variable. Otherwise fine!

**Comment by axic** [comment]
Question is here: https://github.com/ethereum/solidity/pull/6097#issuecomment-467418163 :)

**Comment by axic** [comment]
```
name.sol:2:33: Error: Initial value for constant variable has to be compile-time constant.
  string public constant name = type(C).name;
```

Would need to mark it under `isPure` and add extra code. Can add a syntax test for this.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/codegen/ExpressionCompiler.cpp:1369]: Created an issue #6106

File: Changelog.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ### 0.5.5 (unreleased)
 
 Language Features:
+ * Meta programming: Provide access to the name of contracts.
```
Comment (chriseth) [inline Changelog.md]: Perhaps a short example code how to do it?

**Review by chriseth — APPROVED** [review]

**Pull request merged**